synapse/third_party_rules/access_rules.py

```diff
@@ -0,0 +1,375 @@
+                return False
+
+            # We can only have m.room.member events here. The rule in this case is to only
+            # allow the event if its target is one of the initial two members in the room,
+            # i.e. the state key of one of the two m.room.member states in the room.
+            target = event.state_key
+            for e in member_events:
+                if e.state_key == target:
+                    return True
+
+            return False
+
+        # We're alone in the room (and always have been) and there's one 3PID invite in
+        # flight.
+        if len(member_events) == 1 and len(threepid_invite_events) == 1:
+            # We can only have m.room.member events here. In this case, we can only allow
+            # the event if it's either a m.room.member from the joined user (we can assume
+            # that the only m.room.member event is a join otherwise we wouldn't be able to
+            # send an event to the room) or an an invite event which target is the invited
+            # user.
+            target = event.state_key
+            is_from_threepid_invite = self._is_invite_from_threepid(
+                event, threepid_invite_events[0],
+
+            )
+            if is_from_threepid_invite or target == member_events[0].state_key:
+                return True
+
+            return False
+
+        return True
+
+    @staticmethod
+    def _get_rule_from_state(state_events):
+        """Extract the rule to be applied from the given set of state events.
+
+        Args:
+            state_events (dict[tuple[event type, state key], EventBase]): The set of state
+                events
+        Returns:
+            str, the name of the rule (either "direct", "restricted" or "unrestricted")
+        """
+        access_rules = state_events.get((ACCESS_RULES_TYPE, ""))
+        if access_rules is None:
+            rule = ACCESS_RULE_RESTRICTED
+        else:
+            rule = access_rules.content.get("rule")
+        return rule
+
+    @staticmethod
+    def _is_invite_from_threepid(invite, threepid_invite):
+        """Checks whether the given invite follows the given 3PID invite.
+
+        Args:
+             invite (EventBase): The m.room.member event with "invite" membership.
+             threepid_invite (EventBase): The m.room.third_party_invite event.
+        """
+        token = invite.content.get("third_party_signed", {}).get("token", "")
+        return token == threepid_invite.state_key
+
+    @staticmethod
+    def _is_rule_name_valid(rule):
+        """Returns whether the given rule name is within the allowed values ("direct",
+        "restricted" or "unrestricted").
+
+        Args:
+            rule (str): The name of the rule.
+        Returns:
+            bool, True if the name is valid, False otherwise.
+        """
+        return (
+            rule == ACCESS_RULE_DIRECT
+            or rule == ACCESS_RULE_RESTRICTED
+            or rule == ACCESS_RULE_UNRESTRICTED
```

Comment:
I'd just have a global list called `KNOWN_ACCESS_RULES = (ACCESS_RULE_DIRECT, ....)`, then you can just do `rule in KNOWN_ACCESS_RULES` rather than having a dedicated function.